marimo/_save/cache.py

```diff
@@ -132,30 +131,45 @@ def restore(self, scope: dict[str, Any]) -> None:
                 )
 
     def _restore_from_stub_if_needed(
-        self, value: Any, scope: dict[str, Any]
+        self, value: Any, scope: dict[str, Any], memo: dict[int, Any] | None = None
     ) -> Any:
         """Restore objects from stubs if needed, recursively handling collections."""
+        if memo is None:
+            memo = {}
+        
+        # Check for cycles
+        obj_id = id(value)
+        if obj_id in memo:
+            return memo[obj_id]
+        
         if isinstance(value, ModuleStub):
-            return value.load()
+            result = value.load()
         elif isinstance(value, FunctionStub):
-            return value.load(scope)
+            result = value.load(scope)
         elif isinstance(value, UIElementStub):
-            return value.load()
-        elif isinstance(value, (list, tuple)):
-            # Recursively restore collections
-            restored = [
-                self._restore_from_stub_if_needed(item, scope)
+            result = value.load()
+        elif isinstance(value, list):
+            result = []
```

Comment:
see dict case below

result = [
                self._restore_from_stub_if_needed(item, scope, memo)
                for item in value
            ]
value.clear()
value.extend(result)